### PR TITLE
fix(distributor-ha): init cache on startup

### PR DESCRIPTION
**Pull request by NickAnge**
#### What this PR does
This PR adds the initialization of the local cache of the ha_tracker

The investigation if this is needed has been done in a separate PR https://github.com/grafana/mimir/pull/9853

Check [conclusion](https://github.com/grafana/mimir/pull/9853#issue-2643760579) for TLDR

#### Which issue(s) this PR fixes or relates to

Relates with: https://github.com/grafana/mimir/issues/5796

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/distributor/ha_tracker.go
Comment (aldernero) [inline pkg/distributor/ha_tracker.go:280]: nice, putting this in a separate function makes the purpose of the `loop` much clearer 👍🏻

File: pkg/distributor/ha_tracker.go
Comment (aldernero) [inline pkg/distributor/ha_tracker.go:300]: same comment as above, nice!

**Review by aldernero — APPROVED** [review]
LGTM

File: pkg/distributor/ha_tracker.go
Comment (NickAnge) [reply pkg/distributor/ha_tracker.go:280]: Thank you. Most of this code came from this [PR](https://github.com/grafana/mimir/pull/6991/files). Just addressed comments and refactored some code 🙌

File: pkg/distributor/ha_tracker.go
Comment (NickAnge) [inline pkg/distributor/ha_tracker.go:226]: After running the local setup in [microservices mode ](https://github.com/grafana/mimir/tree/main/development/mimir-microservices-mode), I noticed that this particular piece of code makes the ruler, and queriers fail because of connection issue to consul. 

**Why ?** 
This happens because both `querier` and `ruler` have `distributor` as a dependency in the code. Which means that they initialize the service in both scenarios. 

**Before - init cache**
Before the initialization it was also  failing but when the distributor was in Running state which didn't make the ruler, querier initialization fail.

**After - Init Cache**
Now, the error occurs earlier when the distributor is in the Starting state, causing both ruler and querier to fail.

**How we can solve it ?**
1. Override the mimir.yaml for ruler, and querier to not initialize the ha_tracker : ✅ 
2. Provide connection to consul, which creates confusion since we can see in metrics the ha_tracker from both ruler and querier, thats why I rejected it ❌

File: pkg/distributor/ha_tracker.go
Comment (dimitarvdimitrov) [inline pkg/distributor/ha_tracker.go:241]: do we want to fail startup here or just continue? The `WatchKey` would ignore an unparsable value. Swallowing the error would also make us resilient to random values in the KV store (maybe it's being shared between multiple applications and someone forgot to configure a prefix for Mimir)

File: pkg/distributor/ha_tracker.go
Comment (dimitarvdimitrov) [inline pkg/distributor/ha_tracker.go:247]: again on adding the implications to the log line. This is less clear, but maybe mentioning that the value in the key is skipped should be enough.

File: pkg/distributor/ha_tracker_test.go
Comment (dimitarvdimitrov) [inline pkg/distributor/ha_tracker_test.go:75]: nitpick to make these `const`s. Some editor color variables and constants differently for example

File: pkg/distributor/ha_tracker_test.go
Comment (dimitarvdimitrov) [inline pkg/distributor/ha_tracker_test.go:106]: the update timeouts are different. Is this intentional?

File: pkg/distributor/ha_tracker_test.go
Comment (dimitarvdimitrov) [inline pkg/distributor/ha_tracker_test.go:73]: just double-checking, but this test passes on `main` too. You know that, right? My understanding is that it's just asserting that the tracker works on distributor (re)starts without losing state, not the code you just added.

Can you think of a way to test the new code too? Maybe somehow making sure that given a running HA tracker and querying an already cached replica, the HA tracker does not query the KV store again upon checkReplica?

**Review by dimitarvdimitrov — COMMENTED** [review]
Good job on the investigation! This looks like an improvement we could use 👍 I left mostly small suggestions and questions

File: development/mimir-microservices-mode/config/mimir_override.yaml
Comment (NickAnge) [reply development/mimir-microservices-mode/config/mimir_override.yaml:1]: No, its not by mistake. This `mimir_override.yaml`, is used in rulers and queries in order to disable the ha_tracker. Reasoning explained [here](https://github.com/grafana/mimir/pull/9826#discussion_r1840176310)

I updated also the jsonnet file, which is the source of truth for the docker-compoase.yaml  ([088b23c](https://github.com/grafana/mimir/pull/9826/commits/088b23c1976d225e3654258284198286bdf9332d)) 

Let me know if it makes sense

File: pkg/distributor/ha_tracker.go
Comment (NickAnge) [reply pkg/distributor/ha_tracker.go]: Thank you. Removed it at [253c56a](https://github.com/grafana/mimir/pull/9826/commits/253c56a2f0ce5cc09dcb48cc1eec2273eabdcf3a)

File: pkg/distributor/ha_tracker.go
Comment (NickAnge) [reply pkg/distributor/ha_tracker.go]: Yeah makes sense. I have updated it at [253c56a](https://github.com/grafana/mimir/pull/9826/commits/253c56a2f0ce5cc09dcb48cc1eec2273eabdcf3a). Please let me know what you. think

File: pkg/distributor/ha_tracker.go
Comment (NickAnge) [reply pkg/distributor/ha_tracker.go]: Yeah makes sense. I have updated it at [253c56a](https://github.com/grafana/mimir/pull/9826/commits/253c56a2f0ce5cc09dcb48cc1eec2273eabdcf3a). Please let me know what you. think

File: pkg/distributor/ha_tracker.go
Comment (NickAnge) [reply pkg/distributor/ha_tracker.go:242]: My thought here was that if we have received the key, we should have a respective value, otherwise we should fail 🤔 

However I think it makes sense to continue and not fail here for resilient purposes as you mentioned. I will updated it

File: pkg/distributor/ha_tracker_test.go
Comment (NickAnge) [reply pkg/distributor/ha_tracker_test.go:73]: Yes, the test passes on the main branch as well. You are  right.

I'm struggling to find a way to isolate and test just this function 🤔. When a new ha_tracker is created, the sync* function gets called automatically, initializing the cache. Without this function, checkReplica would still update the cache.

This is why I mentioned that the addition doesn’t add much value for initializing the cache. It justs cause initialization to happen earlier. 

I'll keep looking for alternatives, but any suggestions are very welcome!

File: pkg/distributor/ha_tracker_test.go
Comment (NickAnge) [reply pkg/distributor/ha_tracker_test.go]: Fixed [2e808ec](https://github.com/grafana/mimir/pull/9826/commits/2e808ec33b59399794bf624b2c16ea04fb8b0cf5). Thank you

File: pkg/distributor/ha_tracker_test.go
Comment (NickAnge) [reply pkg/distributor/ha_tracker_test.go]: Fixed [2e808ec](https://github.com/grafana/mimir/pull/9826/commits/2e808ec33b59399794bf624b2c16ea04fb8b0cf5). Thank you

File: development/mimir-microservices-mode/config/mimir_override.yaml
Comment (dimitarvdimitrov) [reply development/mimir-microservices-mode/config/mimir_override.yaml:1]: that's perfect 👌

File: pkg/distributor/ha_tracker_test.go
Comment (dimitarvdimitrov) [reply pkg/distributor/ha_tracker_test.go:73]: some ideas below, _but i also don't want to block this PR on it. You can merge this and then follow-up with more tests or continue the discussion_

you can implement a decorator `kv.Client` and pass that the KV store instead of the consul one here: 

https://github.com/grafana/mimir/blob/c7e1819b5b5c622ad50317a06e091f1c20116f09/pkg/distributor/ha_tracker_test.go#L88


in your implementation you can delegate to a consul client to do the heavy-lifting of a KV-store, but note when calls are being made. For example, you can assert that after startup, no calls to `Get` or `CAS` are made (within the update timeout for example).

That way you assert that the warmup during startup was effective and we didn't go to the KV store because of a cold cache.

File: pkg/distributor/ha_tracker_test.go
Comment (dimitarvdimitrov) [inline pkg/distributor/ha_tracker_test.go:85]: we don't need the preifxed client, do we?

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM, feel free to merge if you want while we keep thinking about the test. Nice work 💪

File: pkg/distributor/ha_tracker_test.go
Comment (NickAnge) [reply pkg/distributor/ha_tracker_test.go]: You are right. Fixed it [2e20498](https://github.com/grafana/mimir/pull/9826/commits/2e204988d140eefe7d1459758b85afffc4803882)

Comment (NickAnge) [reply pkg/distributor/ha_tracker_test.go:85]: I am trying to create a mock client and followed the approach from other tests. However might not needed 🤔 . When I remove it though tests are failing. This code might be replaced at the next PR in favor of [this](https://github.com/grafana/mimir/pull/9826#discussion_r1841194320) .

File: pkg/distributor/ha_tracker_test.go
Comment (NickAnge) [reply pkg/distributor/ha_tracker_test.go:73]: Related PR: https://github.com/grafana/mimir/pull/9905

File: pkg/distributor/ha_tracker_test.go
Comment (NickAnge) [reply pkg/distributor/ha_tracker_test.go:85]: Related PR: https://github.com/grafana/mimir/pull/9905

**Pull request merged**
